pkg/controller/util/endpoint/rate_limiting_queue.go

```diff
@@ -0,0 +1,114 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoint
+
+import (
+	"sync"
+
+	"k8s.io/client-go/util/workqueue"
+)
+
+// RateLimitingQueue implements a wrapper on workqueue.RateLimitingInterface that also rate limits initial key additions
+// to the queue (unlike workqueue.RateLimitingInterface which rate limits only error retries).
+// TODO(mborsz): Rename to something that doesn't collide with workqueue.RateLimitingInterface (or refactor workqueue.RateLimitingInterface).
+// TODO(mborsz): Promote to workqueue package once we validate API in endpoint controllers.
+type RateLimitingQueue struct {
+	queue       workqueue.RateLimitingInterface
+	rateLimiter workqueue.RateLimiter
+
+	queuedNotStarted     map[interface{}]bool
+	queuedNotStartedLock sync.Mutex
+}
+
+// NewRateLimitingQueue creates a new RateLimitingQueue instance.
+func NewRateLimitingQueue(queue workqueue.RateLimitingInterface, rateLimiter workqueue.RateLimiter) *RateLimitingQueue {
+	return &RateLimitingQueue{
+		queue:            queue,
+		rateLimiter:      rateLimiter,
+		queuedNotStarted: make(map[interface{}]bool),
+	}
+}
+
+// Enqueue adds key to the queue if it's not there yet.
+func (r *RateLimitingQueue) Enqueue(key interface{}, allowRateLimiting bool) {
+	r.queuedNotStartedLock.Lock()
+	defer r.queuedNotStartedLock.Unlock()
+	if _, ok := r.queuedNotStarted[key]; ok {
+		// The key is waiting in queue.
+		// There is no point in readding the key, so let's save some rate limiter tokens.
+		return
+	}
+	r.queuedNotStarted[key] = true
+	if allowRateLimiting {
+		// We consume rate limiter's token only if we really add element to the queue.
+		r.queue.AddAfter(key, r.rateLimiter.When(key))
+	} else {
+		r.queue.Add(key)
+	}
+}
+
+// EnqueueRetry adds key back to the queue after error occured while handling.
+// Please note that adding back is not subject of rate limiting at RateLimitingQueue level -- it will
+// be rate limited by the queue itself (potentially using different rate limiter).
+func (r *RateLimitingQueue) EnqueueRetry(key interface{}) {
+	r.queue.AddRateLimited(key)
+}
+
+// Get blocks until the next element is available or queue is shutdown.
+func (r *RateLimitingQueue) Get() (interface{}, bool) {
+	key, quit := r.queue.Get()
+
+	r.queuedNotStartedLock.Lock()
+	defer r.queuedNotStartedLock.Unlock()
+
+	delete(r.queuedNotStarted, key)
+	return key, quit
+}
+
+// DeleteKey marks that given key has been deleted from the system completely and we should free resources associated with it.
+func (r *RateLimitingQueue) DeleteKey(key interface{}) {
+	r.rateLimiter.Forget(key)
```

Comment:
don't you need to call `r.queue.DeleteKey(key)`?

Reply:
also `delete(queuedNotStarted, key)`?

Reply:
Thanks. Done!